### DISCUSSION TITLE
Welcome to fdtracker Discussions!

### DISCUSSION THREAD

**Discussion by mkrzywonski** [mkrzywonski/fdtracker#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by mmurphy001** [comment]
Nice program, looks good and it seems to work so far.

1. What is the logic around screen size/button size/etc... It looks like it is created to be used on a phone or something. -- not a criticism, just asking to the thought pattern around layout. Maybe a small screen such as on a Raspberry Pi? I still have not decided on using a dedicated Pi or old laptop in my FD area.

2. This does not seem to take into consideration the weight of the tray. Are you thinking one should input tare weight? 
    Calculate tare for each tray before inputting it into the program?
I was thinking of putting a bar code on all my trays that way I could shuffle them around in the FD without having to keep track of where they started, also, that would allow me to get a weight of each tray stored so I would not have to deal with tare weight. I'm also eyeing a counting scale with RS232 output. I don't know how that would work but it would make things more efficient. Many years ago I did a bit of work with RS232 and RS422 devices. I've probably forgotten more about it than I originally knew. 

3. Could you explain the "remote" website a bit? I have a home Linux server that has a web server running on it already. having this available outside the input device would be nice. 

4. Is there a way to get the data out of the DB? I already have a MySQL server I use for other purposes with backups/etc.. already setup.

5. I started down the path of creating my own program to do this same thing before thinking of looking for one already in progress. Sorry, I can't help much, I'm a shell script person. Thanks for making this available. Hopefully, this deluge of questions is not more than you bargained for.

**Reply by mkrzywonski** [reply]
1) I wrote it to be used on a phone. I don't think my wife would use anything that required her to set up her laptop every time she runs a batch, and it would take up space she would want to use for prep. Everything is sized to be easily visible on a phone and easy to click with your thumbs.

2) I added "Empty Tray Weight" to the new batch screen. Thanks for the suggestion.

3) Not sure what you're asking about here. I run it on a Raspberry Pi Zero W2, but it should work just fine on your Linux server as well.

4) I may add MySQL support. I will have to refactor the backup/snapshot code to support it.

**Reply by mkrzywonski** [reply]
I added support for MySQL. If you already have production data in the sqlite database, run a git fetch to get the latest version and restart the app, then Go to Tools-->Download Backup. It will download a .zip file to your computer. Edit config.ini (see example.ini) and set up your SQL server, then restart the app. Once you get the app working on MySQL with an empty database, go to Tools-->Restore Backup and upload the .zip file from earlier. Your data will be restored in the MySQL database.

**Reply by mmurphy001** [reply]
I got the mysql option working. The app failed to start 3 times before it worked. I assume it was creating tables/etc... in the failed startup attempts. Thanks for the setup guide, I did not see that until after I got it all working. The systemd setup will help.

**Reply by mkrzywonski** [reply]
Sounds like I need to do some more testing. I tried to build in a database wizard to help with the process of setting up the mysql database. It should have identified conditions like invalid credentials and prompted you with steps to create the database and grant permissions, etc. But, if the app was failing, I must have missed some failure conditions that should be handled. Any info about specific errors or steps to reproduce them would be helpful.

**Comment by mmurphy001** [comment]
I've honestly been trying to figure this out on my own, I've done a couple python/tkinker online courses/books.  But after years of procedural programming (bash scripts only) I guess I'm just not getting it. I got a scale that will output to a PC. I got that working just fine. The problem is the output has spaces before and after the number along with a "g" at the end of the number. such as these examples:

```
  792.0g 
 1560.5g 
  791.5g 
```
From what my testing shows on simple python programs is I could add these two lines to app.py and it should work but it is just not changing anything. The input data seems to be immune to manipulation. It's like I'm not doing anything. Here's what I have:

In "def add_batch():"  I added the last two lines that should have removed the white space and the "g". I know any validation changes would have to be in many more places.
```
           starting_weight = request.form.get(f"starting_weight_{i}", "")
           starting_weight = starting_weight.strip()
           starting_weight = starting_weight.replace("g", "")

```
The error message is "Please enter a number" the only place I can find that phrase is in: ./venv/lib/python3.12/site-packages/pip/_vendor/rich/prompt.py under the float definition. That code is a bit more complex and I was unsuccessful doing the validation there either. And yes I know I should not have been messing with it in the first place.
 
My last idea was to somehow wrap the request.form.get inside the strip and replace but my attempts did not produce results.
Do you have ideas as to how to "validate/manipulate" the input data to just a number by striping out spaces and that pesky "g"?

Thanks,

**Reply by mkrzywonski** [reply]
Nice, your scale sounds like a cool gadget.
Take a look at templates/add_batch.html. This is the template that is used to render the Add Batch page. look at line 23:
`<input type="number" step="0.01" name="starting_weight_${i}" class="form-control" value="${tray.starting_weight}" required>`
The `type="number"` part is what is tripping you up. Your browser is enforcing that requirement on the form input. You can change that to `type="text"` and it will accept the text input and pass it through to your python code that strips out the spaces and removes the g. However, you are going to have to look at templates/view_batch.html as well. There is a section in that template for  in-progress batches, and that section allows you to update the weight for a dry-check. you would have to make a similar change there. That is complicated by the fact that the page calculates the weight loss since the previous measurement. The code for that is Javascript near the bottom of the template file. The javascript runs in the browser before you submit the form, so your python code can't do anything here. You can remove the Javascript code that does this calculation in the browser, or add javascipt to dynamically strip the non-numeric characters from the field value before making the calculation.

**Reply by mkrzywonski** [reply]
I've had a few minutes to ponder this while getting ready for work this morning. I think the best solution is to add a bit of Javascript to both templates to strip out any non-numeric characters from the weight fields. Should be a pretty easy fix and avoids having to deal with the 'type' problems we will have doing math on string values. I'll take a look at it when I get home from work this evening and modify the code.

**Reply by mkrzywonski** [reply]
i have added Javascript to the add_batch and view_batch templates to remove non-numeric characters from the weight input fields. This should allow you to use your scale as an input device without any problems. Let me know how it works for you.

**Reply by mkrzywonski** [reply]
If you like the scale, I'd be curious to know what model it is. Maybe I should buy one for my wife! ;-)

**Comment by mmurphy001** [comment]
I've been a little distracted lately and have not had a chance to look at
the updates yet.
Sorry for the long name but I don't see a part number so here's the
description: "VEVOR Industrial Counting Scale, 15 kg x 0.5 g, with RS232
Port. It is about $80. from vevor.com. I've only had it a few weeks, so far
it has been great, We have a Blue Alpine Freeze Dryer (large size) which
uses the 1/2 sheet pans. I can't read my kitchen scale without putting a
quart paint can on top and balancing the pan on that. Now I can read the
scale directly. Also it is backlit which helps a lot and it plugs in. I"ve
replaced the button cell batteries in my old scale a couple times already.
I've got a pretty good sized bench to work on so space is not a big deal.
This thing does take up a bit of real estate. Getting it set up to pump
data into the PC was way easier than I was expecting. It is almost plug n
play. The directions are a bit fuzzy, look for the stapled on extra
directions in the back of the book. but once I figured it out it was a
breeze. It acts like a keyboard wedge and types the weight every time you
put something on the scale.
Thanks for updating the program. I'll give it a try in a day or two.
